source/session.ts

```diff
@@ -669,28 +670,33 @@ export class Session {
    *   Return update or create promise.
    */
 
-  ensure<T extends Data = Data>(
+  ensure<
+    T extends
+      EntityTypeMap[keyof EntityTypeMap] = EntityTypeMap[keyof EntityTypeMap],
+  >(
     entityType: string,
     data: T,
     identifyingKeys: Array<keyof T> = [],
   ): Promise<T & Entity> {
```

Comment:
I'm thinking if we should make even bigger changes to the generics, in that we start using the entityType as the base of all the generics, to be able to automatically infer it from the entityType used.
So something like
```suggestion
    K extends
      keyof EntityTypeMap = keyof EntityTypeMap,
  >(
    entityType: K,
    data: EntityTypeMap[K],
    identifyingKeys: Array<keyof EntityTypeMap[K]> = [],
  ): Promise<EntityTypeMap[K] & Entity> {
```

It will make the typing stronger but will require more work in changing for people who have already started typing the calls in their applications.

What do you think @ffMathy, @jimmycallin, @lucaas?

Reply:
I love it! Nice work!

Reply:
Waiting for @jimmycallin and @lucaas comments before making the change.

Reply:
Friendly poke (cc @jimmycallin and @lucaas) 🙏 ❤️

Reply:
Decided against this after internal discussions

Reply:
We've been discussing it at the office and it's moving along. There are some things to figure out, we have some new query types that need to fit with this somehow and I need to find the time to do some testing and changes concerning that.

Reply:
Great - keep me posted. Thanks for the update!